File: java/src/main/java/sparkmc/Wizard.java
package sparkmc;

import sparkmc.model.Core;
import sparkmc.model.FlagPreset;
import sparkmc.model.LoaderChannel;
import sparkmc.model.Prepared;
import sparkmc.model.ServerConfig;
import sparkmc.net.Providers;
import sparkmc.util.Ansi;
import sparkmc.util.Reporter;
import sparkmc.util.Util;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.List;
import java.util.Locale;

public final class Wizard {
    private static final BufferedReader IN =
            new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));

    private Wizard() {}

    public static void run(Path dir) {
        banner();
        Core core = pickCore();
        if (core == null) {
            return;
        }
        List<String> versions;
        try {
            info("Loading " + core.label() + " versions...");
            versions = Providers.of(core).versions();
        } catch (Exception e) {
            err(e.getMessage());
            waitEnter();
            return;
        }
        if (versions.isEmpty()) {
            err("no versions returned");
            waitEnter();
            return;
        }
        String version = pickVersion(versions);
        if (version == null) {
            return;
        }
        LoaderChannel channel = null;
        if (core.supportsChannel()) {
            channel = pickChannel();
            if (channel == null) {
                return;
            }
        }
        FlagPreset preset = pickFlags();
        if (preset == null) {
            return;
        }
        String customFlags = "";
        if (preset == FlagPreset.Custom) {
            customFlags = readLine("Custom JVM flags");
            if (customFlags == null || customFlags.isBlank()) {
                err("custom flags cannot be empty");
                waitEnter();
                return;
            }
        }
        Integer ram = pickRam();
        if (ram == null) {
            return;
        }
        boolean noGui = confirm("No GUI (--nogui)?", true);
        boolean autoRestart = confirm("Auto restart on crash?", true);

        ServerConfig cfg = new ServerConfig(core, version, channel, preset, customFlags, ram, noGui, autoRestart);
        System.out.println();
        info("Review");
        row("core", cfg.core().label());
        row("version", cfg.version());
        if (cfg.channel() != null) {
            row("channel", cfg.channel().label());
        }
        row("flags", cfg.preset().label());
        row("ram", cfg.ramMb() + " MB");
        row("nogui", cfg.noGui() ? "yes" : "no");
        row("restart", cfg.autoRestart() ? "yes" : "no");
        System.out.println();
        if (!confirm("Launch setup now?", true)) {
            info("Cancelled");
            return;
        }

        Reporter rep = new Reporter();
        try {
            Prepared prepared = Setup.run(dir, cfg, rep);
            LaunchPlan plan = LaunchPlan.fromConfig(cfg, prepared, dir);
            plan.save(dir);
            info("Configuration saved. Starting server console...");
            ConsoleApp.run(dir);
        } catch (Exception e) {
            err(e.getMessage());
            waitEnter();
        }
    }

    private static void banner() {
        System.out.println(Ansi.GREEN + "sparkmc" + Ansi.RESET + Ansi.GRAY + " - Minecraft server setup" + Ansi.RESET);
        System.out.println(Ansi.GRAY + "Console wizard. Ctrl+C to abort." + Ansi.RESET);
        System.out.println();
    }

    private static Core pickCore() {
        Core[] items = Core.all();
        String[] labels = new String[items.length];
        for (int i = 0; i < items.length; i++) {
            labels[i] = items[i].label();
        }
        Integer idx = pickIndex("Select core", labels, 0);
        return idx == null ? null : items[idx];
    }

    private static LoaderChannel pickChannel() {
        LoaderChannel[] items = LoaderChannel.all();
        String[] labels = new String[items.length];
        for (int i = 0; i < items.length; i++) {
            labels[i] = items[i].label();
        }
        Integer idx = pickIndex("Forge/NeoForge channel", labels, 0);
        return idx == null ? null : items[idx];
    }

    private static FlagPreset pickFlags() {
        FlagPreset[] items = FlagPreset.all();
        String[] labels = new String[items.length];
        for (int i = 0; i < items.length; i++) {
            labels[i] = items[i].label();
        }
        Integer idx = pickIndex("JVM flags preset", labels, 0);
        return idx == null ? null : items[idx];
    }

    private static String pickVersion(List<String> versions) {
        String def = versions.get(0);
        System.out.println(Ansi.CYAN + "Versions:" + Ansi.RESET + " " + Ansi.WHITE + String.join(", ", versions) + Ansi.RESET);
        System.out.println(Ansi.GRAY + "Type the version itself, e.g. 1.12.2 (not a number)." + Ansi.RESET);
        while (true) {
            System.out.println(Ansi.YELLOW + "[sparkmc]" + Ansi.RESET + " " + Ansi.WHITE + "version" + Ansi.RESET
                    + " " + Ansi.GRAY + "[" + def + "]" + Ansi.RESET);
            prompt();
            String input = readRaw();
            if (input == null) {
                return null;
            }
            String chosen = input.isBlank() ? def : input.trim();
            String resolved = resolveVersion(versions, chosen);
            if (resolved != null) {
                info("selected " + resolved);
                return resolved;
            }
            err("unknown version '" + chosen + "'. Example: 1.12.2");
        }
    }

    private static String resolveVersion(List<String> versions, String input) {
        for (String v : versions) {
            if (v.equals(input)) {
                return v;
            }
        }
        String lower = input.toLowerCase(Locale.ROOT);
        String exact = null;
        for (String v : versions) {
            if (v.toLowerCase(Locale.ROOT).equals(lower)) {
                if (exact != null) {
                    return null;
                }
                exact = v;
            }
        }
        if (exact != null) {
            return exact;
        }
        String prefix = null;
        for (String v : versions) {
            if (v.toLowerCase(Locale.ROOT).startsWith(lower)) {
                if (prefix != null) {
                    return null;
                }
                prefix = v;
            }
        }
        return prefix;
    }

    private static Integer pickRam() {
        while (true) {
            System.out.println(Ansi.YELLOW + "[sparkmc]" + Ansi.RESET + " " + Ansi.WHITE + "RAM in MB "
                    + Ansi.GRAY + "[4096]" + Ansi.RESET);
            prompt();
            String input = readRaw();
            if (input == null) {
                return null;
            }
            int value;
            if (input.isBlank()) {
                value = 4096;
            } else {
                try {
                    value = Integer.parseInt(input.trim());
                } catch (NumberFormatException e) {
                    err("enter a number between 1024 and 65536");
                    continue;
                }
                if (value < 1024 || value > 65536) {
                    err("enter a number between 1024 and 65536");
                    continue;
                }
            }
            long heap = Util.heapMb(value);
            if (heap < Util.MIN_HEAP_MB) {
                err("heap would be " + heap + " MB (min " + Util.MIN_HEAP_MB + "), pick more RAM");
                continue;
            }
            info("heap ≈ " + heap + " MB");
            return value;
        }
    }

    private static Integer pickIndex(String title, String[] labels, int def) {
        System.out.println(Ansi.CYAN + title + Ansi.RESET);
        for (int i = 0; i < labels.length; i++) {
            String mark = i == def ? ">" : " ";
            System.out.printf("  %s%s%s %s%2d. %s%s%n", Ansi.GRAY, mark, Ansi.RESET, Ansi.WHITE, i + 1, labels[i], Ansi.RESET);
        }
        while (true) {
            System.out.println(Ansi.YELLOW + "[sparkmc]" + Ansi.RESET + " choose 1-" + labels.length + " "
                    + Ansi.GRAY + "[" + (def + 1) + "]" + Ansi.RESET);
            prompt();
            String input = readRaw();
            if (input == null) {
                return null;
            }
            if (input.isBlank()) {
                return def;
            }
            try {
                int n = Integer.parseInt(input.trim());
                if (n >= 1 && n <= labels.length) {
                    return n - 1;
                }
            } catch (NumberFormatException ignored) {
            }
            err("enter a number from 1 to " + labels.length);
        }
    }

    private static boolean confirm(String promptText, boolean defaultYes) {
        String hint = defaultYes ? "Y/n" : "y/N";
        System.out.println(Ansi.YELLOW + "[sparkmc]" + Ansi.RESET + " " + Ansi.WHITE + promptText + Ansi.RESET
                + " " + Ansi.GRAY + "[" + hint + "]" + Ansi.RESET);
        prompt();
        String input = readRaw();
        if (input == null) {
            return defaultYes;
        }
        if (input.isBlank()) {
            return defaultYes;
        }
        String a = input.trim().toLowerCase(Locale.ROOT);
        return a.startsWith("y") || a.startsWith("д");
    }

    private static String readLine(String promptText) {
        System.out.println(Ansi.YELLOW + "[sparkmc]" + Ansi.RESET + " " + Ansi.WHITE + promptText + Ansi.RESET);
        prompt();
        return readRaw();
    }

    private static void waitEnter() {
        System.out.println(Ansi.GRAY + "press Enter to close" + Ansi.RESET);
        prompt();
        readRaw();
    }

    private static void prompt() {
        System.out.print(Ansi.GREEN + ">" + Ansi.RESET + " ");
        System.out.flush();
    }

    private static String readRaw() {
        try {
            return IN.readLine();
        } catch (Exception e) {
            return null;
        }
    }

    private static void row(String key, String value) {
        System.out.printf("  %s%-10s%s %s%s%s%n", Ansi.GRAY, key, Ansi.RESET, Ansi.WHITE, value, Ansi.RESET);
    }

    private static void info(String msg) {
        System.out.println(Ansi.GREEN + "[sparkmc]" + Ansi.RESET + " " + Ansi.WHITE + msg + Ansi.RESET);
    }

    private static void err(String msg) {
        System.out.println(Ansi.RED + "[sparkmc]" + Ansi.RESET + " " + msg);
    }
}

File: rust/src/wizard.rs
use std::io::{self, BufRead, Write};
use std::path::Path;

use crate::event::Reporter;
use crate::model::{Core, FlagPreset, LoaderChannel, ServerConfig};
use crate::plan::LaunchPlan;
use crate::{net, setup, util};

const RESET: &str = "\x1b[0m";
const WHITE: &str = "\x1b[38;2;216;222;233m";
const GRAY: &str = "\x1b[38;2;118;124;138m";
const GREEN: &str = "\x1b[38;2;138;214;140m";
const RED: &str = "\x1b[38;2;232;118;110m";
const YELLOW: &str = "\x1b[38;2;232;192;102m";
const CYAN: &str = "\x1b[38;2;122;190;238m";

pub fn run(dir: &Path) {
    banner();

    let core = match pick_core() {
        Some(c) => c,
        None => return,
    };

    let versions = match load_versions(core) {
        Ok(v) => v,
        Err(e) => {
            err(&e);
            wait_enter();
            return;
        }
    };
    if versions.is_empty() {
        err("no versions returned");
        wait_enter();
        return;
    }

    let version = match pick_version(&versions) {
        Some(v) => v,
        None => return,
    };

    let channel = if core.supports_channel() {
        match pick_channel() {
            Some(c) => Some(c),
            None => return,
        }
    } else {
        None
    };

    let preset = match pick_flags() {
        Some(p) => p,
        None => return,
    };

    let custom_flags = if matches!(preset, FlagPreset::Custom) {
        match read_line("Custom JVM flags") {
            Some(s) if !s.trim().is_empty() => s,
            _ => {
                err("custom flags cannot be empty");
                wait_enter();
                return;
            }
        }
    } else {
        String::new()
    };

    let ram_mb = match pick_ram() {
        Some(r) => r,
        None => return,
    };

    let no_gui = confirm("No GUI (--nogui)?", true);
    let auto_restart = confirm("Auto restart on crash?", true);

    let cfg = ServerConfig {
        core,
        version,
        channel,
        preset,
        custom_flags,
        ram_mb,
        no_gui,
        auto_restart,
    };

    println!();
    info("Review");
    row("core", cfg.core.label());
    row("version", &cfg.version);
    if let Some(ch) = cfg.channel {
        row("channel", ch.label());
    }
    row("flags", cfg.preset.label());
    row("ram", &format!("{ram_mb} MB"));
    row("nogui", yes_no(cfg.no_gui));
    row("restart", yes_no(cfg.auto_restart));
    println!();

    if !confirm("Launch setup now?", true) {
        info("Cancelled");
        return;
    }

    let rep = Reporter::new();
    let prepared = match setup::run(dir, &cfg, &rep) {
        Ok(p) => p,
        Err(e) => {
            err(&e);
            wait_enter();
            return;
        }
    };

    let plan = LaunchPlan::from_config(&cfg, &prepared, dir);
    if let Err(e) = plan.save(dir) {
        err(&e);
        wait_enter();
        return;
    }

    info("Configuration saved. Starting server console...");
    crate::console::run(dir);
}

fn banner() {
    println!("{GREEN}sparkmc{RESET} {GRAY}- Minecraft server setup{RESET}");
    println!("{GRAY}Console wizard. Ctrl+C to abort.{RESET}");
    println!();
}

fn pick_core() -> Option<Core> {
    let items = Core::all();
    let labels: Vec<&str> = items.iter().map(|c| c.label()).collect();
    pick_index("Select core", &labels, 0).map(|i| items[i])
}

fn prompt_prefix() {
    print!("{GREEN}>{RESET} ");
    let _ = io::stdout().flush();
}

fn pick_version(versions: &[String]) -> Option<String> {
    let default = versions.first().map(String::as_str).unwrap_or("");
    println!("{CYAN}Versions:{RESET} {WHITE}{}{RESET}", versions.join(", "));
    println!("{GRAY}Type the version itself, e.g. 1.12.2 (not a number).{RESET}");
    loop {
        print!("{YELLOW}[sparkmc]{RESET} {WHITE}version{RESET} {GRAY}[{default}]{RESET}\n");
        prompt_prefix();
        let mut line = String::new();
        if io::stdin().read_line(&mut line).is_err() {
            return None;
        }
        let input = line.trim();
        let chosen = if input.is_empty() {
            default.to_string()
        } else {
            input.to_string()
        };
        match resolve_version(versions, &chosen) {
            Some(v) => {
                info(&format!("selected {v}"));
                return Some(v);
            }
            None => {
                err(&format!(
                    "unknown version '{chosen}'. Example: 1.12.2"
                ));
            }
        }
    }
}

fn resolve_version(versions: &[String], input: &str) -> Option<String> {
    let needle = input.trim();
    if needle.is_empty() {
        return None;
    }
    if let Some(exact) = versions.iter().find(|v| v.as_str() == needle) {
        return Some(exact.clone());
    }
    let lower = needle.to_ascii_lowercase();
    let case_hits: Vec<&String> = versions
        .iter()
        .filter(|v| v.to_ascii_lowercase() == lower)
        .collect();
    if case_hits.len() == 1 {
        return Some(case_hits[0].clone());
    }
    let prefix_hits: Vec<&String> = versions
        .iter()
        .filter(|v| v.to_ascii_lowercase().starts_with(&lower))
        .collect();
    if prefix_hits.len() == 1 {
        return Some(prefix_hits[0].clone());
    }
    None
}

fn pick_channel() -> Option<LoaderChannel> {
    let items = LoaderChannel::all();
    let labels: Vec<&str> = items.iter().map(|c| c.label()).collect();
    pick_index("Forge/NeoForge channel", &labels, 0).map(|i| items[i])
}

fn pick_flags() -> Option<FlagPreset> {
    let items = FlagPreset::all();
    let labels: Vec<&str> = items.iter().map(|p| p.label()).collect();
    pick_index("JVM flags preset", &labels, 0).map(|i| items[i])
}

fn pick_ram() -> Option<u32> {
    loop {
        print!("{YELLOW}[sparkmc]{RESET} {WHITE}RAM in MB {GRAY}[4096]{RESET}\n");
        prompt_prefix();
        let mut line = String::new();
        if io::stdin().read_line(&mut line).is_err() {
            return None;
        }
        let trimmed = line.trim();
        let value = if trimmed.is_empty() {
            4096
        } else {
            match trimmed.parse::<u32>() {
                Ok(v) if (1024..=65536).contains(&v) => v,
                _ => {
                    err("enter a number between 1024 and 65536");
                    continue;
                }
            }
        };
        let heap = util::heap_mb(value);
        if heap < util::MIN_HEAP_MB {
            err(&format!(
                "heap would be {heap} MB (min {}), pick more RAM",
                util::MIN_HEAP_MB
            ));
            continue;
        }
        info(&format!("heap ≈ {heap} MB"));
        return Some(value);
    }
}

fn load_versions(core: Core) -> Result<Vec<String>, String> {
    info(&format!("Loading {} versions...", core.label()));
    net::provider(core).versions()
}

fn pick_index(title: &str, labels: &[&str], default: usize) -> Option<usize> {
    println!("{CYAN}{title}{RESET}");
    for (i, label) in labels.iter().enumerate() {
        let mark = if i == default { ">" } else { " " };
        println!("  {GRAY}{mark}{RESET} {WHITE}{:>2}. {label}{RESET}", i + 1);
    }
    loop {
        print!(
            "{YELLOW}[sparkmc]{RESET} choose 1-{} {GRAY}[{}]{RESET}\n",
            labels.len(),
            default + 1
        );
        prompt_prefix();
        let mut line = String::new();
        if io::stdin().read_line(&mut line).is_err() {
            return None;
        }
        let trimmed = line.trim();
        if trimmed.is_empty() {
            return Some(default);
        }
        if let Ok(n) = trimmed.parse::<usize>() {
            if (1..=labels.len()).contains(&n) {
                return Some(n - 1);
            }
        }
        err(&format!("enter a number from 1 to {}", labels.len()));
    }
}

fn confirm(prompt: &str, default_yes: bool) -> bool {
    let hint = if default_yes { "Y/n" } else { "y/N" };
    print!("{YELLOW}[sparkmc]{RESET} {WHITE}{prompt}{RESET} {GRAY}[{hint}]{RESET}\n");
    prompt_prefix();
    let mut line = String::new();
    if io::stdin().read_line(&mut line).is_err() {
        return default_yes;
    }
    let a = line.trim().to_ascii_lowercase();
    if a.is_empty() {
        return default_yes;
    }
    a.starts_with('y') || a.starts_with('д')
}

fn read_line(prompt: &str) -> Option<String> {
    print!("{YELLOW}[sparkmc]{RESET} {WHITE}{prompt}{RESET}\n");
    prompt_prefix();
    let mut line = String::new();
    if io::stdin().lock().read_line(&mut line).is_err() {
        return None;
    }
    Some(line.trim_end_matches(['\r', '\n']).to_string())
}

fn wait_enter() {
    println!("{GRAY}press Enter to close{RESET}");
    prompt_prefix();
    let mut line = String::new();
    let _ = io::stdin().read_line(&mut line);
}

fn row(key: &str, value: &str) {
    println!("  {GRAY}{key:<10}{RESET} {WHITE}{value}{RESET}");
}

fn yes_no(v: bool) -> &'static str {
    if v { "yes" } else { "no" }
}

fn info(msg: &str) {
    println!("{GREEN}[sparkmc]{RESET} {WHITE}{msg}{RESET}");
}

fn err(msg: &str) {
    println!("{RED}[sparkmc]{RESET} {msg}");
}
